### ISSUE TITLE
PbiTools.Deployments.DeploymentException: HTTP Error: Unauthorized

### ISSUE THREAD

**Issue by hcmjensen** [pbi-tools/pbi-tools#229]
Hi, Mathias.  I am able to deploy a dataset into a workspace, but am getting errors when trying to deploy reports.  Here's my manifest (without the actual IDs):

```
{
  "version": "0.10",
  "created": "2022-01-25T17:13:58.5949288+00:00",
  "deployments": {
    "PartnerEUXReport": {
      "description": "All reports deployed from PbixProj sources (.pbix files are not under source control). Dedicated workspace for each environment.",
      "mode": "Report",
      "source": {
        "type": "Folder",
        "path": "./Corp Sales and Bookings/reports/Partner Performance Dashboard_D365"
      },
      "authentication": {
        "type": "ServicePrincipal",
        "tenantId": <tenantid>,
        "clientId": <clientid>,
        "clientSecret": "%PBI_CLIENT_SECRET%"
      },
      "environments": {
        "Development": {
          "workspace": "Corp Sales and Bookings [Development]"
        }
      }
    }
  }
}

```
Since I'm able to deploy a dataset into this workspace, I don't think it's a permissions issue.  Any idea what's going on here?  

Here's the error log:

```
 C:\Users\hannah.jensen\SCM\BIAA_PowerBI> pbi-tools deploy . PartnerEUXReport
Reading PBIXPROJ settings from: C:\Users\hannah.jensen\SCM\BIAA_PowerBI\.pbixproj.json
Starting deployment 'PartnerEUXReport' into environment: Development ...
Found 1 source folders to deploy.
Creating PBIX from report source at: 'C:\Users\hannah.jensen\AppData\Local\Temp\9abd11c1-1850-4163-abde-0addee97a467\Partner Performance Dashboard_D365.pbix'...
Reading PBIXPROJ settings from: C:\Users\hannah.jensen\SCM\BIAA_PowerBI\Corp Sales and Bookings\reports\Partner Performance Dashboard_D365\.pbixproj.json
Using Power BI Desktop install: 2.108.997.0 (22.08) at C:\Program Files\Microsoft Power BI Desktop\bin
Generating PBIX file at 'C:\Users\hannah.jensen\AppData\Local\Temp\9abd11c1-1850-4163-abde-0addee97a467\Partner Performance Dashboard_D365.pbix'...
Access token received. Expires On: 01/14/2023 01:29:34 +00:00
Resolved workspace ID '58d08b12-6f1c-43d5-a071-78845d011407' for workspace: 'Corp Sales and Bookings [Development]'
An unhandled exception occurred.
System.AggregateException: One or more errors occurred. ---> PbiTools.Deployments.DeploymentException: HTTP Error: Unauthorized
{
  "error": {
    "code": "UnknownError",
    "pbi.error": {
      "code": "UnknownError",
      "parameters": {},
      "details": [],
      "exceptionCulprit": 1
    }
  }
} ---> Microsoft.Rest.HttpOperationException: Operation returned an invalid status code 'Unauthorized'
   at Microsoft.PowerBI.Api.ImportsOperations.<UploadFile>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.PowerBI.Api.ImportsOperations.<PostImportFileWithHttpMessage>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.PowerBI.Api.ImportsOperationsExtensions.<PostImportWithFileAsyncInGroup>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PbiTools.Deployments.DeploymentManager.<ImportReportAsync>d__45.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 302    
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at PbiTools.Deployments.DeploymentManager.<DeployReportAsync>d__38.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 99     
   --- End of inner exception stack trace ---
   at PbiTools.Deployments.DeploymentManager.<DeployReportAsync>d__38.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 105    
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PbiTools.Deployments.DeploymentManager.<DeployAsync>d__30.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.cs:line 81
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at PbiTools.Cli.CmdLineActions.Deploy(String folder, String label, String environment, String basePath, Boolean whatIf) in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Cli\Deploy.cs:line 39
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at PowerArgs.ArgAction.Invoke()
   at PbiTools.Program.Main(String[] args) in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Program.cs:line 94
---> (Inner Exception #0) PbiTools.Deployments.DeploymentException: HTTP Error: Unauthorized
{
  "error": {
    "code": "UnknownError",
    "pbi.error": {
      "code": "UnknownError",
      "parameters": {},
      "details": [],
      "exceptionCulprit": 1
    }
  }
} ---> Microsoft.Rest.HttpOperationException: Operation returned an invalid status code 'Unauthorized'
   at Microsoft.PowerBI.Api.ImportsOperations.<UploadFile>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.PowerBI.Api.ImportsOperations.<PostImportFileWithHttpMessage>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.PowerBI.Api.ImportsOperationsExtensions.<PostImportWithFileAsyncInGroup>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PbiTools.Deployments.DeploymentManager.<ImportReportAsync>d__45.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 302    
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at PbiTools.Deployments.DeploymentManager.<DeployReportAsync>d__38.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 99     
   --- End of inner exception stack trace ---
   at PbiTools.Deployments.DeploymentManager.<DeployReportAsync>d__38.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.Report.cs:line 105    
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PbiTools.Deployments.DeploymentManager.<DeployAsync>d__30.MoveNext() in X:\pbi-tools\BRANCHES\1.0.0-rc.2+preview.5\src\PBI-Tools\Deployments\DeploymentManager.cs:line 81<---
```

Thank you very much for your help!
Hannah

**Comment by mthierba** [comment]
You're using the exact same credentials and you're targeting the same workspace as with the dataset deployment you say is working?

**Comment by hcmjensen** [comment]
Correct

**Comment by mthierba** [comment]
It's very difficult to diagnose this with limited information, I'm afraid. The error message says "Unauthorized" which means there's a permissions problem. If you'd like further help, could you describe in more detail the current setup, please?

**Comment by hcmjensen** [comment]
Hi, Mathias.  I figured it out.  The reports were still pointing to a workspace that the service principal didn't have access to, they weren't connected to the dataset I deployed.  Once that was fixed, deployment worked like a charm.  Thank you kindly for your help and for being so responsive.  I really appreciate it!  I'm presenting a proposed pbi-tools workflow to the team next week.  Here's hoping they are receptive to it!
